README.md
# OneWaySMS SDK for Deno

A non-official OneWaySMS SDK written for Deno runtime. Based on the latest [v1.3](http://smsd2.onewaysms.sg/api.pdf) OneWaySMS API documentation.

[![license](https://img.shields.io/github/license/junwen-k/onewaysms-sdk-deno)](https://raw.githubusercontent.com/junwen-k/onewaysms-sdk-deno/master/LICENSE.txt)

## Installing

Import the SDK by using a URL.

```ts
import {
  OneWay,
  OneWayError,
  OneWayErrorType,
} from "https://deno.land/x/gh:junwen-k:onewaysms-sdk-deno/mod.ts";
```

Include the version tag in the URL to install a specific version of the SDK.

```ts
import {
  OneWay,
  OneWayError,
  OneWayErrorType,
} from "https://deno.land/x/gh:junwen-k:onewaysms-sdk-deno@v0.1.0/mod.ts";
```

## Usage and Getting Started

### Initializing a new client

Initialize a new client with the exported `OneWay` class. For instance:

```ts
const svc = new OneWay({
  baseURL: "API_BASE_URL",
  apiUsername: "API_USERNAME",
  apiPassword: "API_PASSWORD",
  senderID: "SENDER_ID",
});
```

### Use case examples

1. **Send SMS** - Send SMS by calling OneWaySMS API gateway, returning mobile terminating ID(s) if request is successful.

   ```ts
   try {
     const data = await svc.sendSMS({
       message: "Hello, 世界",
       mobileNo: ["60123456789", "60129876543"],
     });
     // mtIDs - Mobile terminating IDs
     console.log(data.mtIDs);
   } catch (err) {
     if (err instanceof OneWayError) {
       switch (err.code) {
         case OneWayErrorType.RequestFailure:
         // Handle RequestFailure
         case OneWayErrorType.InvalidCredentials:
         // Handle InvalidCredentials
         case OneWayErrorType.InvalidSenderID:
         // Handle InvalidSenderID
         case OneWayErrorType.InvalidMobileNo:
         // Handle InvalidMobileNo
         case OneWayErrorType.InvalidLanguageType:
         // Handle InvalidLanguageType
         case OneWayErrorType.InvalidMessageCharacters:
         // Handle InvalidMessageCharacters
         case OneWayErrorType.InsufficientCreditBalance:
         // Handle InsufficientCreditBalance
         case OneWayErrorType.UnknownError:
         // Handle UnknownError
         default:
       }
     } else {
       // Handle Generic Error
     }
   }
   ```

1. **Check MT Transaction Status** - Check mobile terminating transaction status based on mobile terminating ID provided. Mobile terminating ID can be obtained by calling send SMS API.

   ```ts
   try {
     const data = await svc.checkTransactionStatus({ mtID: 145712470 });
     switch (data.status) {
       case MTTransactionStatus.Success:
       // Handle success status
       case MTTransactionStatus.TelcoDelivered:
       // Handle telco delivered status
       default:
     }
   } catch (err) {
     if (err instanceof OneWayError) {
       switch (err.code) {
         case OneWayErrorType.MTInvalidNotFound:
         // Handle MTInvalidNotFound
         case OneWayErrorType.MessageDeliveryFailure:
         // Handle MessageDeliveryFailure
         case OneWayErrorType.UnknownError:
         // Handle UnknownError
         default:
       }
     } else {
       // Handle Generic Error
     }
   }
   ```

1. **Check Credit Balance**. Check remaining credit balance for the account in the client's config.

   ```ts
   try {
     const data = await svc.checkCreditBalance();
     // creditBalance - Remaining credit balance for this account
     console.log(data.creditBalance);
   } catch (err) {
     if (err instanceof OneWayError) {
       switch (err.code) {
         case OneWayErrorType.InvalidCredentials:
         // Handle InvalidCredentials
         case OneWayErrorType.UnknownError:
         // Handle UnknownError
         default:
       }
     } else {
       // Handle Generic Error
     }
   }
   ```

## License

This SDK is distributed under the MIT License, see LICENSE.txt for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

- Improve README.md formatting.

### Removed

- `mod.json` and include constant in `oneway.ts` instead.

## [0.1.2] - 2020-06-01

### Added

- `buildRequestURL` function to generate request URL to access OneWaySMS API.
- `deps.ts` to manage dependencies.
- `.gitignore` to ignore vscode settings.

### Updated

- Improve README.md formatting.
- Improve type for request URL params.
- Client to include `User-Agent` header to indicate SDK's request.

## [0.1.1] - 2020-05-29

### Added

- Status code for `OneWayError`.

### Updated

- Tests to include request failure assertion.
- README.md URL example when importing module.

## [0.1.0] - 2020-05-25

- Initial release


oneway_test.ts
// Copyright 2020 KwanJunWen. All rights reserved. MIT license.
import { Status, assertEquals, assert } from "./deps.ts";
import { OneWay } from "./oneway.ts";
import {
  OneWayError,
  OneWayErrorType,
} from "./error.ts";
import { MTTransactionStatus } from "./types.ts";
import { HOSTNAME, PORT } from "./_mock_config.ts";

Deno.test("OneWay:sendSMS -> With valid values", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.sendSMS({
      message: "Hello World",
      mobileNo: "60123456789",
    });
    assertEquals(data.mtIDs, [145712468]);
  } catch (err) {
    assert(!err);
  }
});

Deno.test("OneWay:sendSMS -> With multiple mobileNo", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.sendSMS({
      message: "Hello World",
      mobileNo: ["60123456789", "60129876543"],
    });
    assertEquals(data.mtIDs, [145712468, 145712469]);
  } catch (err) {
    assert(!err);
  }
});

Deno.test("OneWay:sendSMS -> With request failure", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.sendSMS({
      message: "request failure",
      mobileNo: "60123456789",
    });
    assert(!data.mtIDs);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "request failure",
        OneWayErrorType.RequestFailure,
        Status.InternalServerError,
      ),
    );
  }
});

Deno.test("OneWay:sendSMS -> With invalid user credentials", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "invalid",
    apiPassword: "invalid",
    senderID: "SenderID",
  });
  try {
    const data = await svc.sendSMS({
      message: "Hello World",
      mobileNo: "60123456789",
    });
    assert(!data.mtIDs);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "apiusername or apipassword is invalid",
        OneWayErrorType.InvalidCredentials,
      ),
    );
  }
});

Deno.test("OneWay:sendSMS -> With invalid senderID", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "invalid",
  });
  try {
    const data = await svc.sendSMS({
      message: "Hello World",
      mobileNo: "60123456789",
    });
    assert(!data.mtIDs);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "senderid parameter is invalid",
        OneWayErrorType.InvalidSenderID,
      ),
    );
  }
});

Deno.test("OneWay:sendSMS -> With invalid mobileNo", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.sendSMS({
      message: "Hello World",
      mobileNo: "invalid",
    });
    assert(!data.mtIDs);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "mobileno parameter is invalid",
        OneWayErrorType.InvalidMobileNo,
      ),
    );
  }
});

Deno.test("OneWay:sendSMS -> With invalid languageType", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.sendSMS({
      message: "Hello World",
      mobileNo: "60123456789",
      languageType: "2", // Invalid language type in mock server
    });
    assert(!data.mtIDs);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "languagetype is invalid",
        OneWayErrorType.InvalidLanguageType,
      ),
    );
  }
});

Deno.test("OneWay:sendSMS -> With invalid message", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.sendSMS({
      message: "invalid",
      mobileNo: "60123456789",
    });
    assert(!data.mtIDs);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "characters in message are invalid",
        OneWayErrorType.InvalidMessageCharacters,
      ),
    );
  }
});

Deno.test("OneWay:sendSMS -> With insufficient credit balance", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.sendSMS({
      message: "insufficient credit balance",
      mobileNo: "60123456789",
    });
    assert(!data.mtIDs);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "insufficient credit balance",
        OneWayErrorType.InsufficientCreditBalance,
      ),
    );
  }
});

Deno.test("OneWay:sendSMS -> With unknown error", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.sendSMS({
      message: "unknown error",
      mobileNo: "60123456789",
    });
    assert(!data.mtIDs);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "unknown error",
        OneWayErrorType.UnknownError,
      ),
    );
  }
});

Deno.test("OneWay:checkTransactionStatus -> With success status", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.checkTransactionStatus({ mtID: 145712470 });
    assertEquals(data.status, MTTransactionStatus.Success);
  } catch (err) {
    assert(!err);
  }
});

Deno.test("OneWay:checkTransactionStatus -> With telco_delivered status", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.checkTransactionStatus({ mtID: 145712471 });
    assertEquals(data.status, MTTransactionStatus.TelcoDelivered);
  } catch (err) {
    assert(!err);
  }
});

Deno.test("OneWay:checkTransactionStatus -> With invalid mtID", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.checkTransactionStatus({ mtID: 1 });
    assert(!data.status);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "mtid is invalid or not found",
        OneWayErrorType.MTInvalidNotFound,
      ),
    );
  }
});

Deno.test("OneWay:checkTransactionStatus -> With failed mtID", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.checkTransactionStatus({ mtID: 2 });
    assert(!data.status);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "message delivery failed",
        OneWayErrorType.MessageDeliveryFailure,
      ),
    );
  }
});

Deno.test("OneWay:checkTransactionStatus -> With unknown error", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.checkTransactionStatus({ mtID: 3 });
    assert(!data.status);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "unknown error",
        OneWayErrorType.UnknownError,
      ),
    );
  }
});

Deno.test("OneWay:checkCreditBalance -> With valid values", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "Username",
    apiPassword: "Password",
    senderID: "SenderID",
  });
  try {
    const data = await svc.checkCreditBalance();
    assertEquals(data.creditBalance, 6500);
  } catch (err) {
    assert(!err);
  }
});

Deno.test("OneWay:checkCreditBalance -> With invalid user credentials", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "invalid",
    apiPassword: "invalid",
    senderID: "SenderID",
  });
  try {
    const data = await svc.checkCreditBalance();
    assert(!data.creditBalance);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "apiusername or apipassword is invalid",
        OneWayErrorType.InvalidCredentials,
      ),
    );
  }
});

Deno.test("OneWay:checkCreditBalance -> With unknown error", async () => {
  const svc = new OneWay({
    baseURL: `http://${HOSTNAME}:${PORT}`,
    apiUsername: "unknown error",
    apiPassword: "unknown error",
    senderID: "SenderID",
  });
  try {
    const data = await svc.checkCreditBalance();
    assert(!data.creditBalance);
  } catch (err) {
    assertEquals(
      err,
      new OneWayError(
        "unknown error",
        OneWayErrorType.UnknownError,
      ),
    );
  }
});


error.ts
// Copyright 2020 KwanJunWen. All rights reserved. MIT license.
import {
  Status,
  STATUS_TEXT,
} from "./deps.ts";
/**
 * OneWay specific error. Switch based on code to handle specific error
 * when using OneWayClient.
 *
 *    try {
 *      ...
 *    } catch (err) {
 *      if (err instanceof OneWayError) {
 *        switch (err.code) {
 *          case OneWayErrorType.RequestFailure:
 *            // Handle RequestFailure
 *          ...
 *          default:
 *        }
 *      } else {
 *        // Handle Generic Error
 *      }
 *    }
 */
export class OneWayError extends Error {
  readonly code: OneWayErrorType;
  readonly statusCode: Status;

  constructor(message: string, code: OneWayErrorType, statusCode?: Status) {
    const _statusCode = statusCode || Status.OK;
    super(`${statusCode} (${STATUS_TEXT.get(_statusCode)}): ${message}`);
    this.statusCode = _statusCode;
    this.name = "OneWayError";
    this.code = code;
  }
}

export enum OneWayErrorType {
  /**
   * Request Failure error. Error is thrown when response status is not OK.
   */
  RequestFailure = "RequestFailure",
  /**
   * Invalid Credentials error. Error is thrown when API Username or API Password
   * is invalid.
   */
  InvalidCredentials = "InvalidCredentials",
  /**
   * Invalid Sender ID error. Error is thrown when Sender ID is not valid.
   */
  InvalidSenderID = "InvalidSenderID",
  /**
   * Invalid Mobile Number error. Error is thrown when Mobile No is not valid.
   */
  InvalidMobileNo = "InvalidMobileNo",
  /**
   * Invalid LanguageType error. Error is thrown when Language type is not valid.
   */
  InvalidLanguageType = "InvalidLanguageType",
  /**
   * Invalid Message Characters error. Error is thrown when there are invalid characters
   * in the request message.
   */
  InvalidMessageCharacters = "InvalidMessageCharacters",
  /**
   * InsufficientCreditBalance error. Error is thrown when the user does not have
   * sufficient credit balance to perform certain tasks.
   */
  InsufficientCreditBalance = "InsufficientCreditBalance",
  /**
   * Mobile terminating ID invalid or not found error. Error is thrown when Mobile Terminating
   * ID is not a valid ID or not found.
   */
  MTInvalidNotFound = "MTInvalidNotFound",
  /**
   * Message delivery failure error. Error is thrown when Message has been failed to deliver
   * when calling check transaction status API.
   */
  MessageDeliveryFailure = "MessageDeliveryFailure",
  /**
   * Unknown error. Unknown Response returned from OneWay API Gateway.
   */
  UnknownError = "UnknownError",
}


types.ts
// Copyright 2020 KwanJunWen. All rights reserved. MIT license.
export interface OneWayClient {
  sendSMS(input: SendSMSInput): Promise<SendSMSOutput>;
  checkTransactionStatus(
    input: CheckTransactionStatusInput,
  ): Promise<CheckTransactionStatusOutput>;
  checkCreditBalance(): Promise<CheckCreditBalanceOutput>;
}

/**
 * URL path to access OneWaySMS API.
 * "api.aspx" - MT URL path.
 * "bulktrx.aspx" - Check MT transaction URL path.
 * "bulkcredit.aspx" - Check credit balance URL path.
 */
export type OneWayURLPath = "api.aspx" | "bulktrx.aspx" | "bulkcredit.aspx";

/**
 * OneWaySMS API request URL parameters.
 */
export type OneWayURLParams = Record<
  | "apiusername"
  | "apipassword"
  | "senderid"
  | "mobileno"
  | "languagetype"
  | "message"
  | "mtid",
  string
>;

export interface OneWayClientConfig
  extends OneWayUserCredentials, OneWaySenderID {
  /**
   * BaseURL to onewaysms gateway exluding trailing "/".
   * For example: http://gatewayd2.onewaysms.sg:10002
   */
  baseURL: string;
}

export interface OneWayUserCredentials {
  /**
   * Username for your API. (can be obtain under API section once you have log into your account).
   */
  apiUsername: string;
  /**
   * Password for your API. (can be obtain under API section once you have log into your account).
   */
  apiPassword: string;
}

export interface OneWaySenderID {
  /**
   * Refer to sender. Field can take up to 11 alphanumeric characters.
   */
  senderID: string;
}

/**
 * Language Type of the SMS. Indicating the type of sms sent.
 * "1" - Normal Text message (160 characters as 1 MT)
 * "2" - Unicode Text message (70 characters as 1 MT)
 */
export type LanguageType = "1" | "2";

export type SendSMSURLParams = Pick<
  OneWayURLParams,
  | "apiusername"
  | "apipassword"
  | "senderid"
  | "mobileno"
  | "languagetype"
  | "message"
>;

export interface SendSMSInput extends Partial<OneWaySenderID> {
  /**
   * Language Type of the SMS. Refer to LanguageType for details.
   */
  languageType?: LanguageType;
  /**
   * Content of the SMS.
   *
   * Example, if languagetype set to 1, and message
   * has content with length 200, 2 SMS will be sent.
   *
   * SMS can be combined. 7 characters will be used to combine.
   * Below are the character count:
   *
   * 1 SMS = 160 characters
   * 2 SMS = 306 characters (14 characters for joining)
   * 3 SMS = 459 characters (21 characters for joining)
   *
   * We recommend sending up to maximum up to 3 SMS.
   */
  message: string;
  /**
   * Phone number of recipient. Phone number must include country code.
   * For example: 6581234567.
   */
  mobileNo: string | string[];
}

export interface SendSMSOutput {
  /**
   * Mobile terminating ID(s) from the send SMS result.
   */
  mtIDs: number[];
}

export type CheckTransactionStatusURLParams = Pick<OneWayURLParams, "mtid">;

export interface CheckTransactionStatusInput {
  /**
   * Mobile terminating ID returned from the send SMS result.
   */
  mtID: number;
}

export interface CheckTransactionStatusOutput {
  /**
   * Status of the mobile terminating transaction.
   * Refer to MTTransactionStatus for more details.
   */
  status: MTTransactionStatus;
}

export type CheckCreditBalanceURLParams = Pick<
  OneWayURLParams,
  "apiusername" | "apipassword"
>;

export interface CheckCreditBalanceOutput {
  /**
   * Remaining credit balance for the account of this client's config.
   */
  creditBalance: number;
}

export enum MTTransactionStatus {
  /**
   * Successfully sent message.
   */
  Success = "success",
  /**
   * Message has been delivered to Telco.
   */
  TelcoDelivered = "telco_delivered",
}


_mock_server.ts
// Copyright 2020 KwanJunWen. All rights reserved. MIT license.
import {
  listenAndServe,
  ServerRequest,
  Status,
  decode,
} from "./deps.ts";
import { HOSTNAME, PORT } from "./_mock_config.ts";

console.log(`OneWaySMS mock server running at http://${HOSTNAME}:${PORT}/`);
listenAndServe(
  { port: PORT, hostname: HOSTNAME },
  (req: ServerRequest): void => {
    let qs: { [key: string]: string | string[] } = {};
    const qsmatch = req.url.match(/\?(.*)/);
    if (qsmatch && qsmatch.length > 0) {
      qs = decode(qsmatch[1]);
    }
    const {
      apiusername,
      apipassword,
      senderid,
      mobileno,
      message,
      languagetype,
      mtid,
    } = qs;
    if (req.url.includes("/api.aspx")) {
      if (apiusername === "invalid" || apipassword === "invalid") {
        req.respond({ body: "-100" });
        return;
      }
      if (senderid === "invalid") {
        req.respond({ body: "-200" });
        return;
      }
      if (mobileno === "invalid") {
        req.respond({ body: "-300" });
        return;
      }
      // Mock invalid language type
      if (languagetype === "2") {
        req.respond({ body: "-400" });
        return;
      }
      if (message === "invalid") {
        req.respond({ body: "-500" });
        return;
      }
      if (message === "insufficient credit balance") {
        req.respond({ body: "-600" });
        return;
      }
      if (message === "unknown error") {
        req.respond({ body: "random" });
        return;
      }
      if (message === "request failure") {
        req.respond({ status: Status.InternalServerError });
        return;
      }
      if (mobileno === "60123456789,60129876543") {
        req.respond({ body: "145712468,145712469" });
        return;
      }
      if (mobileno === "60123456789") {
        req.respond({ body: "145712468" });
        return;
      }
      return;
    }
    if (req.url.includes("/bulktrx.aspx")) {
      if (mtid === "1") {
        req.respond({ body: "-100" });
        return;
      }
      if (mtid === "2") {
        req.respond({ body: "-200" });
        return;
      }
      if (mtid === "3") {
        req.respond({ body: "random" });
        return;
      }
      if (mtid === "145712470") {
        req.respond({ body: "0" });
        return;
      }
      if (mtid === "145712471") {
        req.respond({ body: "100" });
        return;
      }
      return;
    }
    if (req.url.includes("/bulkcredit.aspx")) {
      if (apiusername === "invalid" || apipassword === "invalid") {
        req.respond({ body: "-100" });
        return;
      }
      if (apiusername === "unknown error" || apipassword === "unknown error") {
        req.respond({ body: "random" });
        return;
      }
      if (apiusername === "Username" || apipassword === "Password") {
        req.respond({ body: "6500" });
        return;
      }
      return;
    }
  },
);


oneway.ts
// Copyright 2020 KwanJunWen. All rights reserved. MIT license.
import {
  CheckCreditBalanceOutput,
  CheckCreditBalanceURLParams,
  CheckTransactionStatusInput,
  CheckTransactionStatusOutput,
  CheckTransactionStatusURLParams,
  LanguageType,
  MTTransactionStatus,
  OneWayClient,
  OneWayClientConfig,
  OneWayURLParams,
  OneWayURLPath,
  SendSMSInput,
  SendSMSOutput,
  SendSMSURLParams,
} from "./types.ts";
import {
  OneWayError,
  OneWayErrorType,
} from "./error.ts";

/**
 * SDK Version constant.
 */
const VERSION = "0.1.2";

// Based on specifications found in http://smsd2.onewaysms.sg/api.pdf.
export class OneWay implements OneWayClient {
  private readonly config: OneWayClientConfig;

  constructor(config: OneWayClientConfig) {
    this.config = config;
  }

  /**
   * Convert message into hexadecimal value.
   * Automatically pad each character to 4 values with "0".
   */
  private messageToHex(message: string): string {
    const buf = [];
    for (let n = 0, l = message.length; n < l; n++) {
      buf.push(Number(message.charCodeAt(n)).toString(16).padStart(4, "0"));
    }
    return buf.join("");
  }

  /**
   * Get language type based on the message provided.
   * Checking is done based on byte length of each character.
   */
  private getLanguageType(message: string): LanguageType {
    const encoder = new TextEncoder();
    for (let n = 0, l = message.length; n < l; n++) {
      if (encoder.encode(message.charAt(n)).byteLength > 1) {
        return "2";
      }
    }
    return "1";
  }

  /**
   * Builds request URL to access OneWaySMS API.
   */
  private buildRequestURL<T = OneWayURLParams>(
    path: OneWayURLPath,
    urlParams: { [P in keyof T]: string },
  ): string {
    const { baseURL } = this.config;
    const params = new URLSearchParams();
    for (const key in urlParams) {
      params.set(key, urlParams[key]);
    }
    return `${baseURL}/${path}?${params.toString()}`;
  }

  /**
   * Builds request URL to send mobile terminating message based on SMS provided.
   */
  private buildSendSMSRequestURL(input: SendSMSInput): string {
    const { apiUsername, apiPassword, senderID } = this.config;

    if (!input.languageType) {
      input.languageType = this.getLanguageType(input.message);
    }

    return this.buildRequestURL<SendSMSURLParams>("api.aspx", {
      apiusername: apiUsername,
      apipassword: apiPassword,
      senderid: input.senderID || senderID,
      mobileno: input.mobileNo.toString(),
      languagetype: input.languageType,
      message: input.languageType === "2"
        ? this.messageToHex(input.message)
        : input.message,
    });
  }

  /**
   * Builds request URL to check transaction status based on mobile terminating ID provided.
   */
  private buildCheckTransactionStatusRequestURL(
    input: CheckTransactionStatusInput,
  ): string {
    return this.buildRequestURL<CheckTransactionStatusURLParams>(
      "bulktrx.aspx",
      { mtid: input.mtID.toString() },
    );
  }

  /**
   * Builds request URL to check remaining credit balance based on API Username and API Password
   * from client's config.
   */
  private buildCheckCreditBalanceRequestURL(): string {
    const { apiUsername, apiPassword } = this.config;

    return this.buildRequestURL<CheckCreditBalanceURLParams>(
      "bulkcredit.aspx",
      { apiusername: apiUsername, apipassword: apiPassword },
    );
  }

  /**
   * Calls "GET" request with User-Agent set to this client library.
   */
  private async getRequest(requestURL: string): Promise<Response> {
    const headers = new Headers();
    headers.set("User-Agent", `onewaysms-sdk-deno/${VERSION}`);
    return fetch(requestURL, {
      headers,
    });
  }

  /**
   * Initiate send SMS request. SMS's language type will be automatically set unless it is
   * defined in the SMS request structure.
   *
   *    const svc = new OneWay({
   *      baseURL: "API_BASE_URL",
   *      apiUsername: "API_USERNAME",
   *      apiPassword: "API_PASSWORD",
   *      senderID: "SENDER_ID",
   *    });
   *    try {
   *      const data = await svc.sendSMS({
   *        message: "Hello, 世界",
   *        mobileNo: ["60123456789", "60129876543"],
   *      });
   *      // mtIDs - Mobile terminating IDs
   *      console.log(data.mtIDs);
   *    } catch (err) {
   *      if (err instanceof OneWayError) {
   *        switch (err.code) {
   *          case OneWayErrorType.RequestFailure:
   *            // Handle RequestFailure
   *          case OneWayErrorType.InvalidCredentials:
   *            // Handle InvalidCredentials
   *          case OneWayErrorType.InvalidSenderID:
   *            // Handle InvalidSenderID
   *          case OneWayErrorType.InvalidMobileNo:
   *            // Handle InvalidMobileNo
   *          case OneWayErrorType.InvalidLanguageType:
   *            // Handle InvalidLanguageType
   *          case OneWayErrorType.InvalidMessageCharacters:
   *            // Handle InvalidMessageCharacters
   *          case OneWayErrorType.InsufficientCreditBalance:
   *            // Handle InsufficientCreditBalance
   *          case OneWayErrorType.UnknownError:
   *            // Handle UnknownError
   *          default:
   *        }
   *      } else {
   *        // Handle Generic Error
   *      }
   *    }
   */
  async sendSMS(
    input: SendSMSInput,
  ): Promise<SendSMSOutput> {
    const requestURL = this.buildSendSMSRequestURL(input);
    return Promise.resolve(
      new Promise((res, rej) => {
        this.getRequest(requestURL).then((resp) => {
          if (!resp.ok) {
            rej(
              new OneWayError(
                "request failure",
                OneWayErrorType.RequestFailure,
                resp.status,
              ),
            );
          }
          return resp.text();
        }).then((respText: string): void => {
          const mtIDs = respText.split(",").map((t: string) => parseInt(t));
          if (mtIDs.length <= 0) {
            rej(
              new OneWayError(
                "unknown error",
                OneWayErrorType.UnknownError,
              ),
            );
            return;
          }
          if (mtIDs[0] > 0) {
            res({ mtIDs });
            return;
          }
          switch (mtIDs[0]) {
            case -100:
              rej(
                new OneWayError(
                  "apiusername or apipassword is invalid",
                  OneWayErrorType.InvalidCredentials,
                ),
              );
              return;
            case -200:
              rej(
                new OneWayError(
                  "senderid parameter is invalid",
                  OneWayErrorType.InvalidSenderID,
                ),
              );
              return;
            case -300:
              rej(
                new OneWayError(
                  "mobileno parameter is invalid",
                  OneWayErrorType.InvalidMobileNo,
                ),
              );
              return;
            case -400:
              rej(
                new OneWayError(
                  "languagetype is invalid",
                  OneWayErrorType.InvalidLanguageType,
                ),
              );
              return;
            case -500:
              rej(
                new OneWayError(
                  "characters in message are invalid",
                  OneWayErrorType.InvalidMessageCharacters,
                ),
              );
              return;
            case -600:
              rej(
                new OneWayError(
                  "insufficient credit balance",
                  OneWayErrorType.InsufficientCreditBalance,
                ),
              );
              return;
            default:
              rej(
                new OneWayError(
                  "unknown error",
                  OneWayErrorType.UnknownError,
                ),
              );
              return;
          }
        })
          .catch(rej);
      }),
    );
  }

  /**
   * Check transaction status based on mobile terminating ID provided.
   *
   *    const svc = new OneWay({
   *      url: "API_BASE_URL",
   *      apiUsername: "API_USERNAME",
   *      apiPassword: "API_PASSWORD",
   *      senderID: "SENDER_ID",
   *    });
   *    try {
   *      const data = await svc.checkTransactionStatus({ mtID: MT_ID });
   *      switch (data.status) {
   *        case MTTransactionStatus.Success:
   *          // Handle success status
   *        case MTTransactionStatus.TelcoDelivered:
   *          // Handle telco delivered status
   *        default:
   *      }
   *    } catch (err) {
   *      if (err instanceof OneWayError) {
   *        switch (err.code) {
   *          case OneWayErrorType.MTInvalidNotFound:
   *            // Handle MTInvalidNotFound
   *          case OneWayErrorType.MessageDeliveryFailure:
   *            // Handle MessageDeliveryFailure
   *          case OneWayErrorType.UnknownError:
   *            // Handle UnknownError
   *          default:
   *        }
   *      } else {
   *        // Handle Generic Error
   *      }
   *    }
   */
  async checkTransactionStatus(
    input: CheckTransactionStatusInput,
  ): Promise<CheckTransactionStatusOutput> {
    const requestURL = this.buildCheckTransactionStatusRequestURL(input);
    return Promise.resolve(
      new Promise((res, rej) => {
        this.getRequest(requestURL)
          .then((resp) => resp.text())
          .then((respText: string): void => {
            const status = parseInt(respText);
            switch (status) {
              case 0:
                res({ status: MTTransactionStatus.Success });
                return;
              case 100:
                res({ status: MTTransactionStatus.TelcoDelivered });
                return;
              case -100:
                rej(
                  new OneWayError(
                    "mtid is invalid or not found",
                    OneWayErrorType.MTInvalidNotFound,
                  ),
                );
                return;
              case -200:
                rej(
                  new OneWayError(
                    "message delivery failed",
                    OneWayErrorType.MessageDeliveryFailure,
                  ),
                );
                return;
              default:
                rej(
                  new OneWayError(
                    "unknown error",
                    OneWayErrorType.UnknownError,
                  ),
                );
                return;
            }
          })
          .catch(rej);
      }),
    );
  }

  /**
   * Check remaining credit balance based on API Username and Password from client's config.
   *
   *    const svc = new OneWay({
   *      baseURL: "API_BASE_URL",
   *      apiUsername: "API_USERNAME",
   *      apiPassword: "API_PASSWORD",
   *      senderID: "SENDER_ID",
   *    });
   *    try {
   *      const data = await svc.checkCreditBalance();
   *      // creditBalance - Remaining credit balance for this account
   *      console.log(data.creditBalance);
   *    } catch (err) {
   *      if (err instanceof OneWayError) {
   *        switch (err.code) {
   *          case OneWayErrorType.InvalidCredentials:
   *            // Handle InvalidCredentials
   *          case OneWayErrorType.UnknownError:
   *            // Handle UnknownError
   *          default:
   *        }
   *      } else {
   *        // Handle Generic Error
   *      }
   *    }
   */
  async checkCreditBalance(): Promise<CheckCreditBalanceOutput> {
    const requestURL = this.buildCheckCreditBalanceRequestURL();
    return Promise.resolve(
      new Promise((res, rej) => {
        this.getRequest(requestURL)
          .then((resp) => resp.text())
          .then((respText: string): void => {
            const creditBalance = parseInt(respText);
            if (creditBalance >= 0) {
              res({ creditBalance });
              return;
            }
            switch (creditBalance) {
              case -100:
                rej(
                  new OneWayError(
                    "apiusername or apipassword is invalid",
                    OneWayErrorType.InvalidCredentials,
                  ),
                );
                return;
              default:
                rej(
                  new OneWayError(
                    "unknown error",
                    OneWayErrorType.UnknownError,
                  ),
                );
                return;
            }
          })
          .catch(rej);
      }),
    );
  }
}
